src/App.tsx

```diff
@@ -14,11 +14,110 @@ const goodsFromServer: string[] = [
   'Garlic',
 ];
 
-const App: React.FC = () => (
-  <div className="App">
-    <h1>Goods</h1>
-    {goodsFromServer.length}
-  </div>
-);
+interface State {
+  goods: string[],
+  goodsVisible: boolean,
+  isSorted: boolean,
+  isSortLength: boolean,
+}
 
-export default App;
+export class App extends React.Component<{}, State> {
+  state: State = {
+    goods: goodsFromServer,
+    goodsVisible: false,
+    isSorted: false,
+    isSortLength: false,
+  };
+
+  showGoods = () => (this.setState(state => ({
+    goodsVisible: !state.goodsVisible,
+  })));
+
+  reverse = () => {
+    this.setState(state => ({
+      goods: [...state.goods].reverse(),
+    }));
+  };
+
+  sortAlph = () => (this.setState(state => (
+    {
+      goods: [...state.goods]
+        .sort((a, b) => (state.isSorted
+          ? b.localeCompare(a)
+          : a.localeCompare(b))),
+      isSorted: !state.isSorted,
+    }
+  )));
+
+  sortLength = () => (this.setState(state => (
+    {
+      goods: ([...state.goods]
+        .sort((a, b) => (state.isSortLength
+          ? b.length - a.length
+          : a.length - b.length))),
+      isSortLength: !state.isSortLength,
+    }
+
+  )));
+
+  reset = () => (this.setState({
+    goods: goodsFromServer,
+    isSorted: false,
+    isSortLength: false,
+  }));
+
+  render() {
+    const { goods, goodsVisible } = this.state;
+
+    return (
+      <div className="App">
+        <h1 className="App__title">Goods</h1>
+        <button
+          type="button"
+          onClick={this.showGoods}
+          className="App__button-title"
+        >
+          {goodsVisible ? 'Hide Goods' : 'Show Goods'}
+        </button>
+        {goodsVisible
+          && (
+            <div className="App__list">
+              {goods.map(good => <li className="App__list--item" key={good}>{good}</li>)}
```

Comment:
This should be separate component `GoodsList` with `ul` - `li` structure inside.